akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -961,18 +965,115 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount:
     }
   }
 
+  private class AsyncCallbackAsMaterializedValueImpl[T](handler: T ⇒ Unit) extends AsyncCallbackAsMaterializedValue[T] {
+    trait CallbackState
+    case class NotInitialized(list: List[T]) extends CallbackState
+    case object Initialized extends CallbackState
+    case object Stopped extends CallbackState
+
+    val concurrentSet = ConcurrentHashMap.newKeySet[Promise[_]]()
+
+    private[this] val callbackState = new AtomicReference[CallbackState](NotInitialized(Nil))
+
+    private[stage] def onStart(): Unit = this.synchronized {
+      val list = callbackState.getAndSet(Initialized: @unchecked) match {
+        case NotInitialized(l) ⇒ l
+      }
+      list.reverse.foreach(ack ⇒ {
+        ack match {
+          case cp: HasCallbackPromise ⇒ concurrentSet.add(cp.done)
+          case _                      ⇒ //no logic for this case
+        }
+        onAsyncInput(ack)
+      })
+    }
+
+    private[stage] def onStop(): Unit = this.synchronized {
+      callbackState.set(Stopped)
+      concurrentSet.asScala.foreach(_.failure(new IllegalStateException("Stream is stopped. Callback is detached.")))
+      concurrentSet.clear()
+    }
+
+    private def onAsyncInput(event: T) = interpreter.onAsyncInput(GraphStageLogic.this, event, handler.asInstanceOf[Any ⇒ Unit])
+
+    override def invokeWithPromise[C <: T with HasCallbackPromise](event: C): Unit = this.synchronized {
+      import akka.dispatch.ExecutionContexts.{ sameThreadExecutionContext ⇒ ec }
+
+      concurrentSet.add(event.done)
+      event.done.future.andThen { case _ ⇒ concurrentSet.remove(event.done) }(ec)
+
+      callbackState.get() match {
+        case Initialized              ⇒ onAsyncInput(event)
+        case list @ NotInitialized(l) ⇒ callbackState.compareAndSet(list, NotInitialized(event :: l))
+        case Stopped                  ⇒ event.done.tryFailure(new IllegalStateException("Stream is stopped. Callback is detached."))
+      }
+    }
+
+    override def invoke(event: T): Unit = this.synchronized {
+      callbackState.get() match {
+        case Initialized              ⇒ onAsyncInput(event)
```

Comment:
what can we do to minimize the synchronized block? I think `onAsyncInput` can be called outside the synchronized here?